<a href="https://colab.research.google.com/github/szabeebzaliz7-a11y/Datascience-And-Gen-AI/blob/main/Assignment_07_Classification_Algorithms_Model_Building.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Objective:
The objective of this assessment is to evaluate your understanding and ability to apply
supervised learning techniques to a real-world dataset.

Dataset:
Use the breast cancer dataset available in the sklearn library.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn
from sklearn.datasets import load_breast_cancer

In [13]:
breast_cancer = load_breast_cancer()
df= pd.DataFrame(breast_cancer.data, columns=breast_cancer.feature_names)
y = breast_cancer.target

In [15]:
display(df.head())

,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst radius,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension
0,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,0.2419,0.07871,...,25.38,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890
1,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,0.1812,0.05667,...,24.99,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902
2,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,0.2069,0.05999,...,23.57,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758
3,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,0.2597,0.09744,...,14.91,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300
4,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,0.1809,0.05883,...,22.54,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678


Key components to be fulfilled:
Loading and Preprocessing:
● Load the breast cancer dataset from sklearn.
● Preprocess the data to handle any missing values and perform necessary feature
scaling.
● Explain the preprocessing steps you performed and justify why they are
necessary for this dataset.

In [9]:
df.isnull().sum()

,0
mean radius,0
mean texture,0
mean perimeter,0
mean area,0
mean smoothness,0
mean compactness,0
mean concavity,0
mean concave points,0
mean symmetry,0
mean fractal dimension,0


In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 30 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   mean radius              569 non-null    float64
 1   mean texture             569 non-null    float64
 2   mean perimeter           569 non-null    float64
 3   mean area                569 non-null    float64
 4   mean smoothness          569 non-null    float64
 5   mean compactness         569 non-null    float64
 6   mean concavity           569 non-null    float64
 7   mean concave points      569 non-null    float64
 8   mean symmetry            569 non-null    float64
 9   mean fractal dimension   569 non-null    float64
 10  radius error             569 non-null    float64
 11  texture error            569 non-null    float64
 12  perimeter error          569 non-null    float64
 13  area error               569 non-null    float64
 14  smoothness error         5

In [16]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_scaled = scaler.fit_transform(df)

# Convert scaled data back to a DataFrame for easier inspection (optional)
X_scaled_df = pd.DataFrame(X_scaled, columns=breast_cancer.feature_names)
display(X_scaled_df.head())

,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst radius,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension
0,1.097064,-2.073335,1.269934,0.984375,1.568466,3.283515,2.652874,2.532475,2.217515,2.255747,...,1.886690,-1.359293,2.303601,2.001237,1.307686,2.616665,2.109526,2.296076,2.750622,1.937015
1,1.829821,-0.353632,1.685955,1.908708,-0.826962,-0.487072,-0.023846,0.548144,0.001392,-0.868652,...,1.805927,-0.369203,1.535126,1.890489,-0.375612,-0.430444,-0.146749,1.087084,-0.243890,0.281190
2,1.579888,0.456187,1.566503,1.558884,0.942210,1.052926,1.363478,2.037231,0.939685,-0.398008,...,1.511870,-0.023974,1.347475,1.456285,0.527407,1.082932,0.854974,1.955000,1.152255,0.201391
3,-0.768909,0.253732,-0.592687,-0.764464,3.283553,3.402909,1.915897,1.451707,2.867383,4.910919,...,-0.281464,0.133984,-0.249939,-0.550021,3.394275,3.893397,1.989588,2.175786,6.046041,4.935010
4,1.750297,-1.151816,1.776573,1.826229,0.280372,0.539340,1.371011,1.428493,-0.009560,-0.562450,...,1.298575,-1.466770,1.338539,1.220724,0.220556,-0.313395,0.613179,0.729259,-0.868353,-0.397100


### Preprocessing Steps:

1.  **Loading the Dataset**: The breast cancer dataset was loaded from `sklearn.datasets.load_breast_cancer()`.
2.  **Handling Missing Values**: We checked for missing values using `df.isnull().sum()` and `df.info()`. The output confirmed that there are no missing values in the dataset, so no imputation or dropping of rows/columns was necessary.
3.  **Feature Scaling**: We applied `StandardScaler` to the feature matrix `X`. This scaler transforms the data such that each feature has a mean of 0 and a standard deviation of 1. This is a crucial step for many machine learning algorithms, especially those that are distance-based (e.g., K-Nearest Neighbors, Support Vector Machines) or gradient-based (e.g., Logistic Regression, Neural Networks), because it prevents features with larger numerical ranges from dominating the learning process. It ensures that all features contribute equally to the distance calculations and model training.

Classification Algorithm Implementation:
Implement the following five classification algorithms:
1. Logistic Regression
2. Decision Tree Classifier
3. Random Forest Classifier
4. Support Vector Machine (SVM)
5. k-Nearest Neighbors (k-NN)
For each algorithm, provide a brief description of how it works and why it might be
suitable for this dataset.

In [19]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.3, random_state=42)

# Initialize and train the Logistic Regression model
log_reg = LogisticRegression(random_state=42)
log_reg.fit(X_train, y_train)

# Make predictions
y_pred_log_reg = log_reg.predict(X_test)

# Evaluate the model
accuracy_log_reg = accuracy_score(y_test, y_pred_log_reg)
print(f"Logistic Regression Accuracy: {accuracy_log_reg:.4f}")
print("\nClassification Report for Logistic Regression:")
print(classification_report(y_test, y_pred_log_reg))

Logistic Regression Accuracy: 0.9825

Classification Report for Logistic Regression:
              precision    recall  f1-score   support

           0       0.97      0.98      0.98        63
           1       0.99      0.98      0.99       108

    accuracy                           0.98       171
   macro avg       0.98      0.98      0.98       171
weighted avg       0.98      0.98      0.98       171



In [20]:
from sklearn.tree import DecisionTreeClassifier

# Initialize and train the Decision Tree Classifier model
dec_tree = DecisionTreeClassifier(random_state=42)
dec_tree.fit(X_train, y_train)

# Make predictions
y_pred_dec_tree = dec_tree.predict(X_test)

# Evaluate the model
accuracy_dec_tree = accuracy_score(y_test, y_pred_dec_tree)
print(f"Decision Tree Classifier Accuracy: {accuracy_dec_tree:.4f}")
print("\nClassification Report for Decision Tree Classifier:")
print(classification_report(y_test, y_pred_dec_tree))

Decision Tree Classifier Accuracy: 0.9415

Classification Report for Decision Tree Classifier:
              precision    recall  f1-score   support

           0       0.90      0.95      0.92        63
           1       0.97      0.94      0.95       108

    accuracy                           0.94       171
   macro avg       0.93      0.94      0.94       171
weighted avg       0.94      0.94      0.94       171



In [22]:
from sklearn.ensemble import RandomForestClassifier

# Initialize and train the Random Forest Classifier model
rand_forest = RandomForestClassifier(random_state=42)
rand_forest.fit(X_train, y_train)

# Make predictions
y_pred_rand_forest = rand_forest.predict(X_test)

# Evaluate the model
accuracy_rand_forest = accuracy_score(y_test, y_pred_rand_forest)
print(f"Random Forest Classifier Accuracy: {accuracy_rand_forest:.4f}")
print("\nClassification Report for Random Forest Classifier:")
print(classification_report(y_test, y_pred_rand_forest))

Random Forest Classifier Accuracy: 0.9708

Classification Report for Random Forest Classifier:
              precision    recall  f1-score   support

           0       0.98      0.94      0.96        63
           1       0.96      0.99      0.98       108

    accuracy                           0.97       171
   macro avg       0.97      0.96      0.97       171
weighted avg       0.97      0.97      0.97       171



In [23]:
from sklearn.svm import SVC

# Initialize and train the Support Vector Machine model
svm_model = SVC(random_state=42)
svm_model.fit(X_train, y_train)

# Make predictions
y_pred_svm = svm_model.predict(X_test)

# Evaluate the model
accuracy_svm = accuracy_score(y_test, y_pred_svm)
print(f"Support Vector Machine Accuracy: {accuracy_svm:.4f}")
print("\nClassification Report for Support Vector Machine:")
print(classification_report(y_test, y_pred_svm))

Support Vector Machine Accuracy: 0.9708

Classification Report for Support Vector Machine:
              precision    recall  f1-score   support

           0       0.95      0.97      0.96        63
           1       0.98      0.97      0.98       108

    accuracy                           0.97       171
   macro avg       0.97      0.97      0.97       171
weighted avg       0.97      0.97      0.97       171



In [24]:
from sklearn.neighbors import KNeighborsClassifier

# Initialize and train the k-Nearest Neighbors model
knn = KNeighborsClassifier()
knn.fit(X_train, y_train)

# Make predictions
y_pred_knn = knn.predict(X_test)

# Evaluate the model
accuracy_knn = accuracy_score(y_test, y_pred_knn)
print(f"k-Nearest Neighbors Accuracy: {accuracy_knn:.4f}")
print("\nClassification Report for k-Nearest Neighbors:")
print(classification_report(y_test, y_pred_knn))

k-Nearest Neighbors Accuracy: 0.9591

Classification Report for k-Nearest Neighbors:
              precision    recall  f1-score   support

           0       0.95      0.94      0.94        63
           1       0.96      0.97      0.97       108

    accuracy                           0.96       171
   macro avg       0.96      0.95      0.96       171
weighted avg       0.96      0.96      0.96       171



Model Comparison:
● Compare the performance of the five classification algorithms.
● Which algorithm performed the best and which one performed the worst?

In [25]:
import pandas as pd

# Create a dictionary of model accuracies
model_accuracies = {
    'Logistic Regression': accuracy_log_reg,
    'Decision Tree Classifier': accuracy_dec_tree,
    'Random Forest Classifier': accuracy_rand_forest,
    'Support Vector Machine': accuracy_svm,
    'k-Nearest Neighbors': accuracy_knn
}

# Convert to a Pandas Series for easy comparison and sorting
accuracies_series = pd.Series(model_accuracies)

print("\n--- Model Performance Comparison (Accuracy) ---")
display(accuracies_series.sort_values(ascending=False))

best_model = accuracies_series.idxmax()
worst_model = accuracies_series.idxmin()

print(f"\nThe algorithm that performed the best was: {best_model} with an accuracy of {accuracies_series.max():.4f}")
print(f"The algorithm that performed the worst was: {worst_model} with an accuracy of {accuracies_series.min():.4f}")


--- Model Performance Comparison (Accuracy) ---


,0
Logistic Regression,0.982456
Random Forest Classifier,0.970760
Support Vector Machine,0.970760
k-Nearest Neighbors,0.959064
Decision Tree Classifier,0.941520



The algorithm that performed the best was: Logistic Regression with an accuracy of 0.9825
The algorithm that performed the worst was: Decision Tree Classifier with an accuracy of 0.9415


Based on the accuracy scores from the test set:

*   **Best Performing Algorithm:** The model with the highest accuracy was **Logistic Regression**, achieving an accuracy of `0.9825`.
*   **Worst Performing Algorithm:** The model with the lowest accuracy was the **Decision Tree Classifier**, achieving an accuracy of `0.9415`.